### DISCUSSION TITLE
Problem with Jinja2 3.1.2 module 'jinja2' has no attribute 'contextfilter' on docker python:3.11.4-slim Not problem on Windows

### DISCUSSION THREAD

**Discussion by letouriste001** [mkdocs/mkdocs#3353] [Q&A]
I created my documentation on windows platform with this package and dependancies:
Jinja2==3.1.2
mkdocs==1.5.2
mkdocs-material==9.2.3
mkdocs-material-extensions==1.1.1
mkdocs-minify-plugin==0.7.1
mkdocs-pdf-export-plugin==0.5.10
mkdocs-plugin-tags==1.0.2
mkdocs-glightbox==0.3.4
mkdocs-localsearch==0.9.2

I ran this command : python -m mkdocs serve
A documentation was correctly build
I can consult the documentation with my browser

I ported the documentation on docker with dockerfile and docker-compose script. I discovered a problem while generating and executed the container:

AttributeError: module 'jinja2' has no attribute 'contextfilter'

```
FROM python:3.11.4-slim

ENV PORT 8000

ADD docs /srv/docs
ADD theme /srv/theme
COPY mkdocs.yml /srv/mkdocs.yml
COPY Deploiement/start.sh /srv/start.sh

WORKDIR /srv

RUN pip install --upgrade pip
RUN pip install Jinja2==3.1.2
RUN pip install mkdocs==1.5.2
RUN pip install mkdocs-material==9.2.3
RUN pip install mkdocs-material-extensions==1.1.1
RUN pip install mkdocs-minify-plugin==0.7.1
RUN pip install mkdocs-pdf-export-plugin==0.5.10
RUN pip install mkdocs-plugin-tags==1.0.2
RUN pip install mkdocs-glightbox==0.3.4
RUN pip install mkdocs-localsearch==0.9.2

EXPOSE 8000

RUN chmod +x ./start.sh
```

I don't understand why it works on windows and not on linux
I saw a fixed simlar problem on github. I also to tryed to downgrade my version of Jinja2 but i encountred another probleme

Can you already encountred this ?
Have you any suggestion ?

**Comment by pawamoy** [comment]
Try to install all your dependencies in a single `pip install` command. Otherwise everytime you install another dependency, it can install other versions of previously installed ones.

```
RUN pip install --upgrade pip
RUN pip install \
    Jinja2==3.1.2 \
    mkdocs==1.5.2 \
    mkdocs-material==9.2.3 \
    mkdocs-material-extensions==1.1.1 \
    mkdocs-minify-plugin==0.7.1 \
    mkdocs-pdf-export-plugin==0.5.10 \
    mkdocs-plugin-tags==1.0.2 \
    mkdocs-glightbox==0.3.4 \
    mkdocs-localsearch==0.9.2 \
```

Also, it'd be better to let MkDocs tell you what Jinja version is needed. You can simply remove your Jinja2 pin. Unless you need a reproducible environment, but in this case you are missing a lot more dependencies here.

**Comment by Andre601** [comment]
Personal recommendation here is to put these dependencies into a `requirements.txt` file (With the same format as shown by pawamoy, meaning `name==version` format) and use `pip install -r requirements.txt`

In my opinion easiest way to install and update multiple dependencies while also allowing tools like dependabot - should you have it as a public repository - to update it automatically.

And as a small side-note, declaring mkdocs here is kinda redundant, as mkdocs-material already defines it as a required dependency to install alongside it.
Also, why define Jinja2 yourself? Afaik is this also installed on Mkdocs itself?
